### ISSUE TITLE
[InstCombine] Failure to avoid decrement for non-zero input to comparison

### ISSUE THREAD

**Issue by RKSimon** [llvm/llvm-project#57635]
https://alive2.llvm.org/ce/z/sZ6wwS

```llvm
declare void @llvm.assume(i1)

define i1 @src(i8 %x) {
  %z = icmp ne i8 %x, 0
  call void @llvm.assume(i1 %z)
  %i = add i8 %x, -1
  %c = icmp ult i8 %i, 7
  ret i1 %c
}

define i1 @tgt(i8 %x) {
  %c = icmp ule i8 %x, 7
  ret i1 %c
}
```
-->
```llvm
----------------------------------------
define i1 @src(i8 %x) denormal-fp-math=ieee,ieee {
%0:
  %z = icmp ne i8 %x, 0
  assume i1 %z
  %i = add i8 %x, 255
  %c = icmp ult i8 %i, 7
  ret i1 %c
}
=>
define i1 @tgt(i8 %x) denormal-fp-math=ieee,ieee {
%0:
  %c = icmp ule i8 %x, 7
  ret i1 %c
}
Transformation seems to be correct!
```

**Comment by RKSimon** [comment]
Candidate Patch: https://reviews.llvm.org/D134172

**Comment by RKSimon** [comment]
Fixed by 09cb9fdef983719b509f43fe0e77762005265039
